content/en/blog/extend-falco-inputs-with-a-plugin-the-basics.md

```diff
@@ -0,0 +1,706 @@
+
+### The functions and methods
+
+#### `main()`
+
+The `main()` function is mandatory for any `go` program, but because we'll build the `plugin` as a library for the *Falco plugin framework* which is written in `C`, we can let it empty.
+
+```go
+// main is mandatory but empty, because the plugin will be used as C library by Falco plugin framework
+func main() {}
+```
+
+#### `init()`
+
+The `init()` function is used for registering our plugin to the `Falco plugin framework`, as a [`source`](https://falco.org/docs/plugins/#source-plugin) and [`extractor`](https://falco.org/docs/plugins/#extractor-plugin).
+
+```go
+// init function is used for referencing our plugin to the Falco plugin framework
+func init() {
+	p := &DockerPlugin{}
+	extractor.Register(p)
+	source.Register(p)
+}
+```
+
+#### `Info()`
+
+This method is mandatory, and all plugins must respect that. It allows the `Falco plugin framework` to have all intel about the plugin itself:
+
+```go
+// Info displays information of the plugin to Falco plugin framework
+func (dockerPlugin *DockerPlugin) Info() *plugins.Info {
+	return &plugins.Info{
+		ID:                 5,
+		Name:               "docker",
+		Description:        "Docker Events",
+		Contact:            "github.com/falcosecurity/plugins/",
+		Version:            "0.1.0",
+		RequiredAPIVersion: "0.3.0",
+		EventSource:        "docker",
+	}
+}
+```
+
+Here some details:
+* `ID`: must be unique among all plugins, it's used by the framework in captures to know which `plugin` is the `source` of events. It's also important for avoiding collisions if you want to share your plugin in the [registry](https://github.com/falcosecurity/plugins). See [documentation](https://falco.org/docs/plugins/#plugin-event-ids) for more details.
+* `Name`: the name of our plugin, will be used in `plugins` section of  `falco.yaml`
+* `EventSource`: this represents the value we'll set in `Falco` rules for mapping, in our case, all rules we'll set will have `source: docker`
+
+#### `Init()`
+
+This method (:warning: different from the function `init()`) will be the first one called by the *Falco plugin framework*, we use it for setting default values for `DockerPlugin` attributes. In our case, these default values are overridden by the value of `init_config:` from `falco.yaml` config file, see []().
+ 
+```go
+// Init is called by the Falco plugin framework as first entry
+// we use it for setting default configuration values and mapping
+// values from `init_config` (json format for this plugin)
+func (dockerPlugin *DockerPlugin) Init(config string) error {
+	dockerPlugin.FlushInterval = 2
+	err := json.Unmarshal([]byte(config), &dockerPlugin)
```

Comment:
Why not `return json.Unmarshal([]byte(config), &dockerPlugin)` ?

Reply:
Because I didn't get the idea :sweat_smile: